File: xyz.java
import java.util.Scanner;
import java.util.ArrayList;

/**
 * Execution: Java executes top-down with JVM bytecode compilation.
 * All primitives are stored on Stack; all Objects reference Heap memory.
 */
public class xyz {

    public static void main(String[] args) {
        System.out.println("Hello World");

        // ==========================================
        // 1. BASICS OF VARIABLES & CONSTANTS
        // ==========================================
        // Syntax: type variableName = value;
        // Java is statically typed. Variables must be declared before use.
        
        int standardVar = 10;
        
        // Constants: declared with final keyword
        final double PI = 3.14159;
        final int MAX_USERS = 100;
        
        System.out.println("Variable: " + standardVar);
        System.out.println("Constant PI: " + PI);


        // ==========================================
        // 2. BUILT-IN PRIMITIVES
        // ==========================================
        
        // Numbers (Whole Numbers)
        // byte: 1 byte (8 bits), range -128 to 127
        byte varByte = 127;
        
        // short: 2 bytes (16 bits), range -32,768 to 32,767
        short varShort = 32767;
        
        // int: 4 bytes (32 bits), range -2,147,483,648 to 2,147,483,647
        int varInt = 2147483647;
        
        // long: 8 bytes (64 bits), requires 'L' suffix
        long varLong = 9223372036854775807L;
        
        // Floating-Point (Decimals)
        // float: 4 bytes (32 bits), precision ~6-7 decimal digits, requires 'f' suffix
        float varFloat = 3.14f;
        
        // double: 8 bytes (64 bits), precision ~15 decimal digits, default for decimals
        double varDouble = 3.141592653589793;
        
        // Character: 2 bytes (16 bits), single Unicode character, single quotes mandatory
        char varChar = 'A';
        
        // Boolean: 1 byte (conceptually), stores true or false
        boolean varBoolean = true;
        
        System.out.println("\n--- Primitives ---");
        System.out.println("byte: " + varByte);
        System.out.println("short: " + varShort);
        System.out.println("int: " + varInt);
        System.out.println("long: " + varLong);
        System.out.println("float: " + varFloat);
        System.out.println("double: " + varDouble);
        System.out.println("char: " + varChar);
        System.out.println("boolean: " + varBoolean);


        // ==========================================
        // 3. REFERENCE TYPES (OBJECTS)
        // ==========================================
        // Reference types store memory address pointing to Heap objects.
        // Stack holds the reference variable; Heap holds the actual object.
        
        // Strings: immutable objects
        String varStr = "Hello";
        String anotherStr = "World";
        
        // Arrays: ordered collections of same type
        int[] varArrInt = {10, 20, 30, 40};
        String[] varArrStr = new String[3];
        varArrStr[0] = "Alice";
        varArrStr[1] = "Bob";
        varArrStr[2] = "Charlie";
        
        // ArrayList: dynamic array (equivalent to C++ std::vector)
        ArrayList<String> namesList = new ArrayList<>();
        namesList.add("Alice");
        namesList.add("Bob");
        namesList.add("Charlie");
        
        // null: represents absence of value (only for reference types, NOT primitives)
        String nullValue = null;
        
        System.out.println("\n--- Reference Types ---");
        System.out.println("String: " + varStr);
        System.out.println("Array[0]: " + varArrInt[0]);
        System.out.println("ArrayList: " + namesList);
        System.out.println("null value: " + nullValue);


        // ==========================================
        // 4. OPERATORS
        // ==========================================
        
        // Arithmetic operators: +, -, *, /, %
        int a = 10, b = 3;
        System.out.println("\n--- Arithmetic Operators ---");
        System.out.println(a + " + " + b + " = " + (a + b));
        System.out.println(a + " - " + b + " = " + (a - b));
        System.out.println(a + " * " + b + " = " + (a * b));
        System.out.println(a + " / " + b + " = " + (a / b));
        System.out.println(a + " % " + b + " = " + (a % b));
        
        // Comparison operators: ==, !=, >, <, >=, <=
        System.out.println("\n--- Comparison Operators ---");
        System.out.println(a + " > " + b + " is " + (a > b));
        System.out.println(a + " < " + b + " is " + (a < b));
        System.out.println(a + " == " + b + " is " + (a == b));
        
        // Logical operators: && (AND), || (OR), ! (NOT)
        System.out.println("\n--- Logical Operators ---");
        System.out.println("true && false = " + (true && false));
        System.out.println("true || false = " + (true || false));
        System.out.println("!true = " + (!true));
        
        // Assignment operators: =, +=, -=, *=, /=, %=
        int num = 20;
        num += 5;  // num = num + 5
        System.out.println("\nAfter num += 5: " + num);
        
        // Unary operators: ++, --, +, -
        int counter = 0;
        System.out.println("counter: " + counter);
        System.out.println("counter++: " + counter++);  // Post-increment
        System.out.println("++counter: " + (++counter));  // Pre-increment


        // ==========================================
        // 5. CONDITIONALS (If Family)
        // ==========================================
        // if statement evaluates boolean condition
        
        System.out.println("\n--- If Conditionals ---");
        int age = 25;
        
        if (age < 18) {
            System.out.println("You are a minor");
        } else if (age >= 18 && age < 65) {
            System.out.println("You are an adult");
        } else {
            System.out.println("You are a senior");
        }
        
        // Nested conditionals
        int score = 85;
        if (score >= 80) {
            if (score >= 90) {
                System.out.println("Grade: A");
            } else {
                System.out.println("Grade: B");
            }
        } else {
            System.out.println("Grade: C or below");
        }


        // ==========================================
        // 6. SWITCH & TERNARY OPERATORS
        // ==========================================
        
        System.out.println("\n--- Switch Statement ---");
        // switch only works with: byte, short, char, int, String, Enum
        int dayNum = 3;
        String dayName;
        
        switch (dayNum) {
            case 1:
                dayName = "Monday";
                break;
            case 2:
                dayName = "Tuesday";
                break;
            case 3:
                dayName = "Wednesday";
                break;
            default:
                dayName = "Unknown";
        }
        System.out.println("Day " + dayNum + ": " + dayName);
        
        // Ternary operator: (condition) ? valueIfTrue : valueIfFalse
        System.out.println("\n--- Ternary Operator ---");
        int userAge = 20;
        String status = (userAge >= 18) ? "Adult" : "Minor";
        System.out.println("Status: " + status);
        
        // Enhanced Switch (Java 14+)
        System.out.println("\n--- Enhanced Switch Expression ---");
        String month = "March";
        int days = switch (month) {
            case "January", "March", "May", "July", "August", "December" -> 31;
            case "April", "June", "September", "November" -> 30;
            case "February" -> 28;
            default -> 0;
        };
        System.out.println("Days in " + month + ": " + days);


        // ==========================================
        // 7. LOOPS
        // ==========================================
        
        System.out.println("\n--- While Loop ---");
        int count = 0;
        while (count < 3) {
            System.out.println("While iteration: " + count);
            count++;
        }
        
        // Do-while (executes at least once)
        System.out.println("\n--- Do-While Loop ---");
        count = 0;
        do {
            System.out.println("Do-while iteration: " + count);
            count++;
        } while (count < 2);
        
        // Standard for loop
        System.out.println("\n--- For Loop ---");
        for (int i = 0; i < 3; i++) {
            System.out.println("For iteration: " + i);
            if (i == 1) {
                System.out.println("  (Skipping next iteration)");
                continue;
            }
        }
        
        // Enhanced for-each loop (for iterables)
        System.out.println("\n--- For-Each Loop ---");
        int[] numbers = {10, 20, 30, 40};
        for (int num_val : numbers) {
            System.out.println("Array element: " + num_val);
        }
        
        // Loop with break
        System.out.println("\n--- Loop with Break ---");
        for (int i = 0; i < 5; i++) {
            if (i == 3) {
                System.out.println("Breaking at i=" + i);
                break;
            }
            System.out.println("i=" + i);
        }


        // ==========================================
        // 8. INPUT & OUTPUT
        // ==========================================
        
        System.out.println("\n--- Output ---");
        System.out.println("println adds newline");
        System.out.print("print ");
        System.out.print("without newline");
        System.out.println();
        
        // Formatted output with printf
        System.out.println("\n--- Formatted Output (printf) ---");
        int age_out = 25;
        double salary = 50000.50;
        char grade = 'A';
        
        System.out.printf("Age: %d%n", age_out);
        System.out.printf("Salary: $%.2f%n", salary);
        System.out.printf("Grade: %c%n", grade);
        System.out.printf("%-15s %8d%n", "Apple", 25);  // Left-align text, right-align number
        
        // Input with Scanner
        System.out.println("\n--- Scanner Input ---");
        Scanner sc = new Scanner(System.in);
        // Uncomment to test input:
        // System.out.print("Enter your name: ");
        // String inputName = sc.nextLine();
        // System.out.println("Hello, " + inputName);
        
        // Alternative: demonstrate without actual input
        String demoName = "John";
        int demoAge = 30;
        System.out.println("Entered name: " + demoName);
        System.out.println("Entered age: " + demoAge);


        // ==========================================
        // 9. STRING FORMATTING
        // ==========================================
        
        System.out.println("\n--- String Formatting ---");
        double price = 19.9934;
        int population = 1500000;
        
        // String.format() returns formatted string
        String formatted1 = String.format("Price: $%.2f", price);
        System.out.println(formatted1);
        
        // printf directly to console
        System.out.printf("Population: %,d%n", population);
        
        // String concatenation
        String message = "Price: $" + String.format("%.2f", price);
        System.out.println(message);


        // ==========================================
        // 10. PARSING & TYPE CONVERSION
        // ==========================================
        
        System.out.println("\n--- Type Conversion ---");
        
        // String to Integer
        String ageInput = "25";
        int parsedAge = Integer.parseInt(ageInput);
        System.out.println("Parsed int: " + parsedAge);
        
        // String to Double
        String priceInput = "99.99";
        double parsedPrice = Double.parseDouble(priceInput);
        System.out.println("Parsed double: " + parsedPrice);
        
        // Integer to String
        int score_val = 99;
        String scoreString1 = String.valueOf(score_val);  // Recommended
        String scoreString2 = Integer.toString(score_val);
        System.out.println("Number to String: " + scoreString1);
        
        // Explicit casting (narrowing)
        double decimalValue = 10.7;
        int castToInt = (int) decimalValue;  // Truncates to 10
        System.out.println("Cast 10.7 to int: " + castToInt);


        // ==========================================
        // 11 & 12. FUNCTIONS/METHODS & PARAMETERS
        // ==========================================
        
        System.out.println("\n--- Methods/Functions ---");
        
        // Call method that returns a value
        int sum = addNumbers(5, 10);
        System.out.println("5 + 10 = " + sum);
        
        // Pass by value (primitives)
        int originalValue = 10;
        System.out.println("Before passByValue: " + originalValue);
        passByValue(originalValue);
        System.out.println("After passByValue: " + originalValue);  // Still 10
        
        // Pass by reference (objects)
        int[] originalArray = {100};
        System.out.println("Before passByReference: " + originalArray[0]);
        passByReference(originalArray);
        System.out.println("After passByReference: " + originalArray[0]);  // Changed to 999
        
        // Variable arguments (varargs)
        int sumVarargs = sumMultiple(1, 2, 3, 4, 5);
        System.out.println("Sum of 1,2,3,4,5: " + sumVarargs);
        
        // Default parameter behavior (Java doesn't have default params, but overloading)
        printMessage("Hello");  // Uses overloaded version
        printMessage("Hello", "World");
  
        sc.close();  // Close Scanner resource
        System.out.println("\n--- Program Complete ---");
    }


    // ==========================================
    // METHOD DEFINITIONS
    // ==========================================
    
    // Simple method with return value
    public static int addNumbers(int a, int b) {
        return a + b;
    }
    
    // Demonstrate pass by value (primitives)
    // Changes inside method don't affect original
    public static void passByValue(int val) {
        val = 999;
        System.out.println("Inside method (pass by value): " + val);
    }
    
    // Demonstrate pass by reference (objects)
    // Changes to object content affect original
    public static void passByReference(int[] arr) {
        arr[0] = 999;  // Modifies the actual array
        System.out.println("Inside method (pass by reference): " + arr[0]);
    }
    
    // Method with variable number of arguments (varargs)
    public static int sumMultiple(int... numbers) {
        int total = 0;
        for (int num : numbers) {
            total += num;
        }
        return total;
    }
    
    // Method overloading (different parameter lists, same method name)
    public static void printMessage(String msg) {
        System.out.println("Message: " + msg);
    }
    
    public static void printMessage(String msg1, String msg2) {
        System.out.println("Message: " + msg1 + " " + msg2);
    }
}

File: xyz.py
# built-in Python module that provides access to the Python interpreter and system-level information.
import sys 
def main():
    print("Hello World")
    
    # ==========================================
    # 1. BASICS OF VARIABLES & CONSTANTS
    # ==========================================
    # Syntax: variable_name = value
    # Python is dynamically typed; types are inferred at runtime, not declared.
    # Variables act as labels/tags bound to objects in memory, not literal memory buckets.
    
    dynamic_var = 10        # Binds label 'dynamic_var' to integer object 10
    dynamic_var = "String"  # Rebinds label to a string object (perfectly valid)

    # Constants:
    # Python does not have a strict 'const' keyword enforcing immutability.
    # By convention, ALL_CAPS denotes a variable that should not be changed.
    PI = 3.14159
    MAX_CONNECTIONS = 100

    # ==========================================
    # 2. BUILT-IN PRIMITIVES (Strictly speaking, all are Objects in Python)
    # ==========================================
    # Python's "primitives" are immutable objects allocated on the private heap.
    
    # Numbers (Integers)
    # int: Arbitrary-precision integers. Python dynamically allocates more memory (bytes) as the number grows. No risk of overflow
    var_int = 2147483647
    var_large_int = 9223372036854775807123456789  

    # Floating-Point (Decimals)
    # float: Implemented using C doubles (IEEE 754 double precision). ~15 decimal digits. No 'f' or 'd' suffix needed.
    var_float = 3.141592653589793

    # Complex Numbers (Python specific built-in)
    var_complex = 3 + 4j # Real part 3.0, Imaginary part 4.0

    # Character / String (str)
    # Python has NO separate 'char' type. A character is just a string of length 1.
    # UTF-8 encoded Unicode objects. Single '' or double "" quotes are identical.
    var_char = 'A'

    # Boolean (bool)
    # Subclass of int. True behaves like 1, False behaves like 0.
    # Capitalized in Python.
    var_boolean = True 


    # ==========================================
    # 3. REFERENCE TYPES & MEMORY ARCHITECTURE
    # ==========================================
    # Python Memory Model: All objects live on the Heap. Variables live in frames on the Stack 
    # and hold C-style pointers (references) to the Heap objects.
    
    # Immutable Reference Types (Cannot be changed after creation):
    # Strings, Tuples, Ints, Floats. Reassigning creates a *new* object.
    text = "Hello World"
    print(f"Original: {text}")
    print(f"Upper: {text.upper()}")
    print(f"Lower: {text.lower()}")
    print(f"Length: {len(text)}")
    print(f"Split: {text.split()}")
    var_tuple = (10, 20, 30) # Immutable equivalent of a fixed array

    # Mutable Reference Types (Can be changed in-place):
    # Lists, Dictionaries, Sets.
    var_list = [10, 20, 30, 40] # Dynamic array (similar to ArrayList in Java, heavily optimized C-array of pointers under the hood)
    my_list = [10, 20, 30]
    my_list.append(40)
    my_list.extend([50, 60])
    print(f"List after append/extend: {my_list}")
    print(f"List length: {len(my_list)}")
    print(f"List[2]: {my_list[2]}")

    # Dictionaries
    my_dict = {"name": "Alice", "age": 30}
    my_dict["city"] = "NYC"
    print(f"Dictionary: {my_dict}")
    print(f"Keys: {list(my_dict.keys())}")
    print(f"Values: {list(my_dict.values())}")
    # Special Types: None (Equivalent to Java's null)
    # Represents the intentional absence of a value. It is a singleton object of type NoneType.
    var_null = None


    # ==========================================
    # 4. OPERATORS
    # ==========================================
    # Arithmetic:
    # +, -, *, / (Always returns float), % (Modulo)
    # // (Floor Division - truncates decimal), ** (Exponentiation)
    calc = (10 // 3) + (2 ** 3) # Evaluates to 3 + 8 = 11

    # Comparison (Relational):
    # == (Value equality), !=, >, <, >=, <=

    # Logical:
    # Uses English words, NOT symbols (&&, ||, !).
    # 'and', 'or', 'not'
    logic_res = (True and False) or not False

    # Assignment:
    # =, +=, -=, *=, /=, //=, %=, **=
    # NOTE: Python does NOT have ++ or -- (unary increment/decrement). Use += 1 or -= 1.

    # Identity vs Equality & Membership (Python Specific):
    # 'is' checks MEMORY ADDRESS (identity), '==' checks VALUE equality.
    # 'in' checks if an element exists in an iterable.
    list_a = [1, 2]
    list_b = [1, 2]
    print(list_a == list_b) # True (Same value)
    print(list_a is list_b) # False (Different objects in heap)
    print(1 in list_a)      # True


    # ==========================================
    # 5. CONDITIONALS (If Family)
    # ==========================================
    # Uses indentation (whitespace) instead of curly braces {} to define scope blocks.
    # 'else if' is contracted to 'elif'.
    
    condition, another_condition = False, True
    if condition:
        pass # 'pass' is a null operation placeholder
    elif another_condition:
        # Nested block
        if 10 > 5:
            pass
        else:
            pass
    else:
        pass


    # ==========================================
    # 6. Match/Switch
    # ==========================================
    # Structural Pattern Matching (Python 3.10+):
    # Replaces Java's switch. Much more powerful, allows matching object structures, types, and unpacking.
    # No 'break' statement is needed; cases do not fall through.
    status_code = 200
    match status_code:
        case 200 | 201: # OR operator in match
            response = "Success"
        case 400:
            response = "Bad Request"
        case _: # Underscore is the default/wildcard (equivalent to default:)
            response = "Unknown"

    # ============================================
    # 7. Ternary
    # =============================================
    # Ternary Operator:
    # Syntax: [value_if_true] if [condition] else [value_if_false]
    ternary_age = 20
    ternary_status = "Adult" if ternary_age >= 18 else "Minor"


    # ==========================================
    # 7. LOOPS
    # ==========================================
    # While Loop:
    count = 0
    while count < 3:
        count += 1

    # Python does NOT have a do-while loop natively.
    # Simulated do-while:
    while True:
        # Do something
        break # Exit condition checked at the end

    # For Loop (Always a For-Each loop in Python):
    # Python loops over iterables directly. 
    collection = ["A", "B", "C"]
    for element in collection:
        pass

    # For a traditional index-based loop, use range(start, stop_exclusive, step)
    for i in range(0, 5, 1): 
        pass

    # Loop Control: 'break', 'continue', and 'else'
    # Python loops can have an 'else' block which executes ONLY if the loop finishes naturally (no break).
    for num in range(3):
        if num == 5:
            break
    else:
        print("Loop completed without hitting a break statement.")


    # ==========================================
    # 8. INPUT & OUTPUT
    # ==========================================
    # Standard Output
    # print() automatically appends a newline by default.
    print("message")
    
    # To print without a newline, modify the 'end' parameter.
    print("No-Newline!", end="") 
    
    # To print multiple items with a custom separator:
    print("\nItem1", "Item2", sep=" | ")

    # Standard Input
    # input() reads an entire line of text as a String. Standard input buffer handles the \n.
    # (Uncomment below to test input)
    # user_input = input("Enter your name: ") 


    # ==========================================
    # 9. FORMATTING
    # ==========================================
    # 1. f-Strings (Formatted String Literals - Python 3.6+) - Most modern, readable approach.
    price = 19.9934
    population = 1500000
    
    # Evaluates expressions directly inside {}. 
    # :.2f (2 decimal places), :, (comma separator)
    print(f"Price: ${price:.2f}")               # Output: Price: $19.99
    print(f"Population: {population:,}")        # Output: Population: 1,500,000
    
    # Padding and alignment: < (left), > (right), ^ (center)
    print(f"{'Apple':<10} {5:>5}")              # Left-align text 10 chars, right-align num 5 chars

    # 2. .format() method (Older, but still heavily used)
    print("char= {}, int={}, float={:.2f}".format(var_char, var_int, var_float))

    # 3. % formatting (Legacy C-style, identical to Java's printf)
    print("char= %c, int=%d, float=%.2f" % (var_char, var_int, var_float))


    # ==========================================
    # 10. PARSING & TYPE CONVERSION (Casting)
    # ==========================================
    # Python relies on built-in constructor functions to cast/convert between types.
    
    # String to Number (Parsing)
    age_input = "25"
    parsed_age = int(age_input)       # Equivalent to Integer.parseInt()
    price_input = "99.99"
    parsed_float = float("3.14")      # Equivalent to Double.parseDouble()
    print(f"String '{price_input}' to float: {parsed_float}")

    # Number to String
    score_val = 99
    score_str = str(score_val) # Equivalent to String.valueOf()
    print(f"Int {score_val} to string: '{score_str}'")
    
    # Type checking with isinstance()
    print(f"\nType checking:")
    print(f"isinstance(25, int): {isinstance(25, int)}")
    print(f"isinstance('text', str): {isinstance('text', str)}")
    print(f"isinstance([1,2], list): {isinstance([1,2], list)}\n")

    # Type coercion is explicit in Python (except int to float operations).
    # You cannot concatenate str + int without casting the int to str first.
    # e.g., "Age: " + str(25)


# ==========================================
# 11 & 12. FUNCTIONS (Methods)
# ==========================================
# Basics & Where to define:
# Defined using the 'def' keyword. They must be defined BEFORE they are called in the script.
# Python functions are first-class objects (they can be assigned to variables, passed as arguments).

# Defining, Parameters, and Type Hinting (Python 3.5+):
# Syntax: def function_name(param: type) -> return_type:
def modify_data(num: int, arr: list) -> list:
    """
    Pass By Object Reference (Call-by-sharing):
    - Variables point to objects. The function receives a copy of the POINTER, not the object.
    - If the object is Immutable (e.g., int, str, tuple), modifying it rebinds the local 
      pointer to a new object. The original variable outside remains unchanged.
    - If the object is Mutable (e.g., list, dict), modifying it in-place affects the 
      underlying heap object. The original variable outside sees the change.
    """
    # 'num' is an int (immutable). This assignment points the local 'num' to a new int object.
    # Original integer remains completely unaffected.
    num = 999 
    
    # 'arr' is a list (mutable). .append() modifies the object dynamically on the heap.
    # The caller will see this mutation.
    arr.append(4) 
    
    return arr

    # Function with return value
    def add_numbers(a: int, b: int) -> int:
        """Simple function with type hints."""
        return a + b
    
    result = add_numbers(5, 10)
    print(f"add_numbers(5, 10) = {result}")
    
    # Pass by object reference demonstration
    def modify_primitive(num: int) -> None:
        num = 999
        print(f"  Inside function (primitive): {num}")
    
    def modify_list(lst: list) -> None:
        lst.append(999)
        print(f"  Inside function (list): {lst}")
    
    # Primitive (immutable)
    original_int = 10
    print(f"\nBefore modify_primitive: {original_int}")
    modify_primitive(original_int)
    print(f"After modify_primitive: {original_int} (unchanged)")
    
    # List (mutable)
    original_list = [1, 2, 3]
    print(f"\nBefore modify_list: {original_list}")
    modify_list(original_list)
    print(f"After modify_list: {original_list} (modified!)")
    
    # Default parameters
    def greet(name: str, greeting: str = "Hello") -> str:
        return f"{greeting}, {name}!"
    
    print(f"\ngreet('Alice'): {greet('Alice')}")
    print(f"greet('Bob', 'Hi'): {greet('Bob', 'Hi')}")
    
    # Variable length arguments
    def sum_all(*numbers: int) -> int:
        return sum(numbers)

    
# Execution Flow entry point mapping (Similar to public static void main)
# 'sys.argv' contains command line arguments.
if __name__ == "__main__":
    main()
    
    # Demonstrating Pass by Object Reference behavior:
    original_int = 10
    original_list = [1, 2, 3]
    
    modify_data(original_int, original_list)
    # original_int is still 10. original_list is now [1, 2, 3, 4]